ResponseMonitor keeps polling after CDP WebSocket disconnects

## Problem

When a new chat is created in Antigravity while the bot is monitoring a response, the CDP WebSocket connection disconnects. `ResponseMonitor` has no awareness of this disconnection and continues polling every 2 seconds, with every poll failing with "WebSocket is not connected" error, for up to 5 minutes (300s timeout).

## Root Cause

`ResponseMonitor` does not listen to `CdpService`'s `disconnected` / `reconnected` / `reconnectFailed` events. The `poll()` catch block (line ~900) logs the error but lets polling continue indefinitely.

## Expected Behavior

- On CDP disconnect: immediately pause polling, set phase to `'disconnected'`
- On CDP reconnect: resume polling seamlessly
- On CDP reconnect failure: stop monitoring and trigger `onTimeout` callback

## Affected Files

- `src/services/responseMonitor.ts` — main fix target
- `src/services/cdpService.ts` — event source (no changes needed)
- `tests/services/responseMonitor.test.ts` — test additions

## Reproduction

1. Start the bot with `npm run start`
2. Send a message from Discord that triggers Antigravity
3. While Antigravity is generating, create a new chat in Antigravity UI
4. Observe: ResponseMonitor emits "WebSocket is not connected" errors every 2 seconds for 5 minutes

## Logs

```
[9:43:47] [PHASE] Generating (38 chars)
[9:43:51] [ERROR] [CdpConnectionPool] Workspace "LazyGravity" disconnected
[9:43:51] [ERROR] [ResponseMonitor] poll error: Error: WebSocket is not connected
... (repeats every 2 seconds for ~5 minutes)
[9:48:45] [WARN] Timeout (36 chars captured)
```

--- a/src/services/responseMonitor.ts
+++ b/src/services/responseMonitor.ts
@@ -438,7 +438,7 @@ export const RESPONSE_SELECTORS = {
 };
 
 /** Response generation phases */
-export type ResponsePhase = 'waiting' | 'thinking' | 'generating' | 'complete' | 'timeout' | 'quotaReached';
+export type ResponsePhase = 'waiting' | 'thinking' | 'generating' | 'complete' | 'timeout' | 'quotaReached' | 'disconnected';
 
 export interface ResponseMonitorOptions {
     /** CDP service instance */
@@ -484,7 +484,6 @@ export class ResponseMonitor {
     private readonly onProcessLog?: (text: string) => void;
 
     private pollTimer: ReturnType<typeof setTimeout> | null = null;
-    private timeoutTimer: ReturnType<typeof setTimeout> | null = null;
     private isRunning: boolean = false;
     private lastText: string | null = null;
     private baselineText: string | null = null;
@@ -494,6 +493,15 @@ export class ResponseMonitor {
     private quotaDetected: boolean = false;
     private seenProcessLogKeys: Set<string> = new Set();
     private structuredDiagLogged: boolean = false;
+
+    // CDP disconnect handling (#48)
+    private isPaused: boolean = false;
+    private onCdpDisconnected: (() => void) | null = null;
+    private onCdpReconnected: (() => void) | null = null;
+    private onCdpReconnectFailed: ((err: Error) => void | Promise<void>) | null = null;
+
+    // Activity-based timeout (#49)
+    private lastActivityTime: number = 0;
 
     constructor(options: ResponseMonitorOptions) {
         this.cdpService = options.cdpService;
@@ -528,6 +536,7 @@ export class ResponseMonitor {
     private async initMonitoring(passive: boolean): Promise<void> {
         if (this.isRunning) return;
         this.isRunning = true;
+        this.isPaused = false;
         this.lastText = null;
         this.baselineText = null;
         this.generationStarted = passive;
@@ -592,23 +601,15 @@ export class ResponseMonitor {
             }
         }
 
-        // Set timeout timer
-        if (this.maxDurationMs > 0) {
-            this.timeoutTimer = setTimeout(async () => {
-                const lastText = this.lastText ?? '';
-                this.setPhase('timeout', lastText);
-                await this.stop();
-                try {
-                    await Promise.resolve(this.onTimeout?.(lastText));
-                } catch (error) {
-                    logger.error('[ResponseMonitor] timeout callback failed:', error);
-                }
-            }, this.maxDurationMs);
-        }
+        // Activity-based timeout: track last activity time instead of fixed timer (#49)
+        this.lastActivityTime = Date.now();
+
+        // Register CDP connection event listeners (#48)
+        this.registerCdpConnectionListeners();
 
         const mode = passive ? 'Passive monitoring' : 'Monitoring';
         logger.debug(
-            `── ${mode} started | poll=${this.pollIntervalMs}ms timeout=${this.maxDurationMs / 1000}s baseline=${this.baselineText?.length ?? 0}ch`,
+            `── ${mode} started | poll=${this.pollIntervalMs}ms inactivityTimeout=${this.maxDurationMs / 1000}s baseline=${this.baselineText?.length ?? 0}ch`,
         );
 
         this.schedulePoll();
@@ -617,13 +618,11 @@ export class ResponseMonitor {
     /** Stop monitoring */
     async stop(): Promise<void> {
         this.isRunning = false;
+        this.isPaused = false;
+        this.unregisterCdpConnectionListeners();
         if (this.pollTimer) {
             clearTimeout(this.pollTimer);
             this.pollTimer = null;
-        }
-        if (this.timeoutTimer) {
-            clearTimeout(this.timeoutTimer);
-            this.timeoutTimer = null;
         }
     }
 
@@ -686,6 +685,9 @@ export class ResponseMonitor {
                 case 'quotaReached':
                     logger.warn('Quota Reached');
                     break;
+                case 'disconnected':
+                    logger.warn(`CDP Disconnected — paused (${len} chars captured)`);
+                    break;
                 default:
                     logger.phase(`${phase}`);
             }
@@ -693,8 +695,63 @@ export class ResponseMonitor {
         }
     }
 
+    private registerCdpConnectionListeners(): void {
+        this.onCdpDisconnected = () => {
+            if (!this.isRunning) return;
+            logger.warn('[ResponseMonitor] CDP disconnected — pausing poll');
+            this.isPaused = true;
+            if (this.pollTimer) {
+                clearTimeout(this.pollTimer);
+                this.pollTimer = null;
+            }
+            this.setPhase('disconnected', this.lastText);
+        };
+
+        this.onCdpReconnected = () => {
+            if (!this.isRunning) return;
+            logger.warn('[ResponseMonitor] CDP reconnected — resuming poll');
+            this.isPaused = false;
+            this.lastActivityTime = Date.now();
+            const resumePhase = this.generationStarted ? 'generating' : 'waiting';
+            this.setPhase(resumePhase, this.lastText);
+            this.schedulePoll();
+        };
+
+        this.onCdpReconnectFailed = async (err: Error) => {
+            if (!this.isRunning) return;
+            logger.error('[ResponseMonitor] CDP reconnection failed — stopping monitor:', err.message);
+            const lastText = this.lastText ?? '';
+            this.setPhase('disconnected', lastText);
+            await this.stop();
+            try {
+                await Promise.resolve(this.onTimeout?.(lastText));
+            } catch (error) {
+                logger.error('[ResponseMonitor] timeout callback failed:', error);
+            }
+        };
+
+        this.cdpService.on('disconnected', this.onCdpDisconnected);
+        this.cdpService.on('reconnected', this.onCdpReconnected);
+        this.cdpService.on('reconnectFailed', this.onCdpReconnectFailed);
+    }
+
+    private unregisterCdpConnectionListeners(): void {
+        if (this.onCdpDisconnected) {
+            this.cdpService.removeListener('disconnected', this.onCdpDisconnected);
+            this.onCdpDisconnected = null;
+        }
+        if (this.onCdpReconnected) {
+            this.cdpService.removeListener('reconnected', this.onCdpReconnected);
+            this.onCdpReconnected = null;
+        }
+        if (this.onCdpReconnectFailed) {
+            this.cdpService.removeListener('reconnectFailed', this.onCdpReconnectFailed);
+            this.onCdpReconnectFailed = null;
+        }
+    }
+
     private schedulePoll(): void {
-        if (!this.isRunning) return;
+        if (!this.isRunning || this.isPaused) return;
         this.pollTimer = setTimeout(async () => {
             await this.poll();
             if (this.isRunning) {
@@ -730,6 +787,7 @@ export class ResponseMonitor {
             newEntries.push(normalized.slice(0, 300));
         }
         if (newEntries.length > 0) {
+            this.lastActivityTime = Date.now();
             try {
                 this.onProcessLog?.(newEntries.join('\n\n'));
             } catch {
@@ -833,6 +891,7 @@ export class ResponseMonitor {
 
             // Handle stop button appearing
             if (isGenerating) {
+                this.lastActivityTime = Date.now();
                 if (!this.generationStarted) {
                     this.generationStarted = true;
                     this.setPhase('thinking', null);
@@ -870,6 +929,7 @@ export class ResponseMonitor {
             // Text change handling
             const textChanged = effectiveText !== null && effectiveText !== this.lastText;
             if (textChanged) {
+                this.lastActivityTime = Date.now();
                 this.lastText = effectiveText;
 
                 if (this.currentPhase === 'waiting' || this.currentPhase === 'thinking') {
@@ -896,6 +956,19 @@ export class ResponseMonitor {
                     }
                     return;
                 }
+            }
+
+            // Activity-based inactivity timeout (#49)
+            if (this.maxDurationMs > 0 && Date.now() - this.lastActivityTime >= this.maxDurationMs) {
+                const lastText = this.lastText ?? '';
+                this.setPhase('timeout', lastText);
+                await this.stop();
+                try {
+                    await Promise.resolve(this.onTimeout?.(lastText));
+                } catch (error) {
+                    logger.error('[ResponseMonitor] timeout callback failed:', error);
+                }
+                return;
             }
         } catch (error) {
             logger.error('[ResponseMonitor] poll error:', error);

--- a/tests/services/responseMonitor.lean.test.ts
+++ b/tests/services/responseMonitor.lean.test.ts
@@ -9,7 +9,7 @@
  *   - Constructor accepts only: cdpService, pollIntervalMs (default 2000),
  *     maxDurationMs, stopGoneConfirmCount (default 3), onProgress, onComplete,
  *     onTimeout, onPhaseChange
- *   - NO network event subscription (no on/removeListener on cdpService)
+ *   - NO network event subscription, but subscribes to CDP connection events (disconnected/reconnected/reconnectFailed)
  *   - NO onActivity, networkCompleteDelayMs, textStabilityCompleteMs, etc.
  *   - NO getLastExtractionSource(), getLastDomActivityLines()
  *   - Each poll makes exactly 3 CDP calls: stop button, quota, text extraction
@@ -38,11 +38,13 @@ interface LeanResponseMonitorOptions {
     //          noUpdateTimeoutMs, noTextCompletionDelayMs, completionStabilityMs
 }
 
-// Minimal mock: only call, getPrimaryContextId. NO on/removeListener.
+// Minimal mock: call, getPrimaryContextId, on, removeListener (for CDP connection events).
 function createMockCdpService() {
     return {
         call: jest.fn().mockResolvedValue({ result: { value: null } }),
         getPrimaryContextId: jest.fn().mockReturnValue(1),
+        on: jest.fn(),
+        removeListener: jest.fn(),
     };
 }
 
@@ -452,22 +454,30 @@ describe('Lean ResponseMonitor (new API)', () => {
     });
 
     // ---------------------------------------------------------------
-    // Test 11: NO network event subscription (cdpService.on should NOT be called)
+    // Test 11: Subscribes to CDP connection events (not network events)
     // ---------------------------------------------------------------
-    it('does NOT subscribe to network events (no cdpService.on calls)', async () => {
+    it('subscribes to CDP connection events on start and removes them on stop', async () => {
         const monitor = createMonitor();
 
         cdpService.call.mockResolvedValueOnce(cdpResult(null));
         await monitor.start();
 
-        // The new lean monitor should NOT call on() on the cdpService
-        expect(cdpService).not.toHaveProperty('on');
-        // If it has an 'on' mock somehow, ensure it was never called
-        if (typeof (cdpService as any).on === 'function') {
-            expect((cdpService as any).on).not.toHaveBeenCalled();
-        }
+        // Should subscribe to CDP connection events (disconnected, reconnected, reconnectFailed)
+        expect(cdpService.on).toHaveBeenCalledWith('disconnected', expect.any(Function));
+        expect(cdpService.on).toHaveBeenCalledWith('reconnected', expect.any(Function));
+        expect(cdpService.on).toHaveBeenCalledWith('reconnectFailed', expect.any(Function));
+        expect(cdpService.on).toHaveBeenCalledTimes(3);
+
+        // Should NOT subscribe to network events
+        expect(cdpService.on).not.toHaveBeenCalledWith('Network.requestWillBeSent', expect.any(Function));
+        expect(cdpService.on).not.toHaveBeenCalledWith('Network.loadingFinished', expect.any(Function));
 
         await monitor.stop();
+
+        // Should unregister all CDP connection listeners
+        expect(cdpService.removeListener).toHaveBeenCalledWith('disconnected', expect.any(Function));
+        expect(cdpService.removeListener).toHaveBeenCalledWith('reconnected', expect.any(Function));
+        expect(cdpService.removeListener).toHaveBeenCalledWith('reconnectFailed', expect.any(Function));
     });
 
     // ---------------------------------------------------------------
@@ -551,6 +561,184 @@ describe('Lean ResponseMonitor (new API)', () => {
         await jest.advanceTimersByTimeAsync(2000);
 
         expect(cdpService.call.mock.calls.length - callsBefore).toBe(4);
+
+        await monitor.stop();
+    });
+
+    // ---------------------------------------------------------------
+    // CDP disconnect handling tests (#48)
+    // ---------------------------------------------------------------
+
+    it('pauses polling on CDP disconnect and resumes on reconnect', async () => {
+        const onProgress = jest.fn();
+        const onPhaseChange = jest.fn();
+
+        const monitor = createMonitor({ onProgress, onPhaseChange });
+
+        cdpService.call.mockResolvedValueOnce(cdpResult(null)); // baseline
+        await monitor.start();
+
+        // Poll 1: detect generation with text
+        cdpService.call
+            .mockResolvedValueOnce(cdpResult({ isGenerating: true }))
+            .mockResolvedValueOnce(cdpResult(false))
+            .mockResolvedValueOnce(cdpResult('Hello'))
+            .mockResolvedValueOnce(cdpResult([])); // process logs
+        await jest.advanceTimersByTimeAsync(2000);
+        expect(onProgress).toHaveBeenCalledWith('Hello');
+
+        // Simulate CDP disconnect
+        const disconnectHandler = cdpService.on.mock.calls.find(
+            (c: any[]) => c[0] === 'disconnected',
+        )?.[1];
+        expect(disconnectHandler).toBeDefined();
+        disconnectHandler!();
+
+        expect(onPhaseChange).toHaveBeenCalledWith('disconnected', 'Hello');
+
+        // Advance 4 seconds — no polls should fire while paused
+        const callCountAtDisconnect = cdpService.call.mock.calls.length;
+        await jest.advanceTimersByTimeAsync(4000);
+        expect(cdpService.call.mock.calls.length).toBe(callCountAtDisconnect);
+
+        // Simulate CDP reconnect
+        const reconnectHandler = cdpService.on.mock.calls.find(
+            (c: any[]) => c[0] === 'reconnected',
+        )?.[1];
+        expect(reconnectHandler).toBeDefined();
+        reconnectHandler!();
+
+        // Phase should be restored
+        expect(onPhaseChange).toHaveBeenCalledWith('generating', 'Hello');
+
+        // Polling resumes — set up mock for next poll
+        cdpService.call
+            .mockResolvedValueOnce(cdpResult({ isGenerating: true }))
+            .mockResolvedValueOnce(cdpResult(false))
+            .mockResolvedValueOnce(cdpResult('Hello World'))
+            .mockResolvedValueOnce(cdpResult([]));
+        await jest.advanceTimersByTimeAsync(2000);
+        expect(onProgress).toHaveBeenCalledWith('Hello World');
+
+        await monitor.stop();
+    });
+
+    it('calls onTimeout when CDP reconnection fails', async () => {
+        const onTimeout = jest.fn();
+        const monitor = createMonitor({ onTimeout });
+
+        cdpService.call.mockResolvedValueOnce(cdpResult(null)); // baseline
+        await monitor.start();
+
+        // Poll 1: detect text
+        cdpService.call
+            .mockResolvedValueOnce(cdpResult({ isGenerating: true }))
+            .mockResolvedValueOnce(cdpResult(false))
+            .mockResolvedValueOnce(cdpResult('Partial'))
+            .mockResolvedValueOnce(cdpResult([]));
+        await jest.advanceTimersByTimeAsync(2000);
+
+        // Simulate disconnect + reconnect failure
+        const disconnectHandler = cdpService.on.mock.calls.find(
+            (c: any[]) => c[0] === 'disconnected',
+        )?.[1];
+        disconnectHandler!();
+
+        const reconnectFailedHandler = cdpService.on.mock.calls.find(
+            (c: any[]) => c[0] === 'reconnectFailed',
+        )?.[1];
+        reconnectFailedHandler!(new Error('Max retries'));
+
+        // Let the stop() promise resolve
+        await jest.advanceTimersByTimeAsync(0);
+
+        expect(onTimeout).toHaveBeenCalledWith('Partial');
+        expect(monitor.isActive()).toBe(false);
+    });
+
+    // ---------------------------------------------------------------
+    // Activity-based timeout tests (#49)
+    // ---------------------------------------------------------------
+
+    it('does not timeout while text is actively changing', async () => {
+        let timedOutText: string | null = null;
+        const monitor = createMonitor({
+            maxDurationMs: 6000,
+            onTimeout: (text) => { timedOutText = text; },
+        });
+
+        cdpService.call.mockResolvedValueOnce(cdpResult(null)); // baseline
+        await monitor.start();
+
+        // Poll 1 at 2s: text changes → activity resets
+        cdpService.call
+            .mockResolvedValueOnce(cdpResult({ isGenerating: true }))
+            .mockResolvedValueOnce(cdpResult(false))
+            .mockResolvedValueOnce(cdpResult('Line 1'))
+            .mockResolvedValueOnce(cdpResult([]));
+        await jest.advanceTimersByTimeAsync(2000);
+
+        // Poll 2 at 4s: text changes again → activity resets
+        cdpService.call
+            .mockResolvedValueOnce(cdpResult({ isGenerating: true }))
+            .mockResolvedValueOnce(cdpResult(false))
+            .mockResolvedValueOnce(cdpResult('Line 1\nLine 2'))
+            .mockResolvedValueOnce(cdpResult([]));
+        await jest.advanceTimersByTimeAsync(2000);
+
+        // Poll 3 at 6s: text changes again → activity resets
+        cdpService.call
+            .mockResolvedValueOnce(cdpResult({ isGenerating: true }))
+            .mockResolvedValueOnce(cdpResult(false))
+            .mockResolvedValueOnce(cdpResult('Line 1\nLine 2\nLine 3'))
+            .mockResolvedValueOnce(cdpResult([]));
+        await jest.advanceTimersByTimeAsync(2000);
+
+        // 6 seconds have passed (exceeds maxDurationMs), but activity kept resetting
+        expect(timedOutText).toBeNull();
+
+        await monitor.stop();
+    });
+
+    it('activity-based timeout fires only after inactivity, not fixed duration', async () => {
+        let timedOutText: string | null = null;
+        const monitor = createMonitor({
+            maxDurationMs: 4000, // 4s inactivity timeout
+            onTimeout: (text) => { timedOutText = text; },
+        });
+
+        cdpService.call.mockResolvedValueOnce(cdpResult(null)); // baseline
+        await monitor.start();
+
+        // Poll 1 at 2s: text appears → activity resets
+        cdpService.call
+            .mockResolvedValueOnce(cdpResult({ isGenerating: true }))
+            .mockResolvedValueOnce(cdpResult(false))
+            .mockResolvedValueOnce(cdpResult('Line 1'))
+            .mockResolvedValueOnce(cdpResult([]));
+        await jest.advanceTimersByTimeAsync(2000);
+
+        // Poll 2 at 4s: text changes → activity resets (would be fixed timeout here)
+        cdpService.call
+            .mockResolvedValueOnce(cdpResult({ isGenerating: true }))
+            .mockResolvedValueOnce(cdpResult(false))
+            .mockResolvedValueOnce(cdpResult('Line 1\nLine 2'))
+            .mockResolvedValueOnce(cdpResult([]));
+        await jest.advanceTimersByTimeAsync(2000);
+
+        // At 4s: fixed timeout would have fired, but activity-based should NOT
+        expect(timedOutText).toBeNull();
+
+        // Poll 3 at 6s: text changes again → activity resets
+        cdpService.call
+            .mockResolvedValueOnce(cdpResult({ isGenerating: true }))
+            .mockResolvedValueOnce(cdpResult(false))
+            .mockResolvedValueOnce(cdpResult('Line 1\nLine 2\nLine 3'))
+            .mockResolvedValueOnce(cdpResult([]));
+        await jest.advanceTimersByTimeAsync(2000);
+
+        // At 6s: 2x maxDurationMs, but still no timeout thanks to activity
+        expect(timedOutText).toBeNull();
 
         await monitor.stop();
     });

